Memory leaks fix, part I: create/destroy

jodator opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Fix memory leaks. Closes ckeditor/ckeditor5#1341.

---

### Additional information

* This PR introduces a test util for testing memory leaks using mocha and requires changes from ckeditor5-dev repo: https://github.com/ckeditor/ckeditor5-dev/pull/473.
* Alternatively the test can be run on manually spawned Chrome instance and attaching to to a karma runner. To run Google Chrome with required flags execute in command line (other Chrome instances must be closed):
```sh
google-chrome \
    -js-flags="--expose-gc" \
    --enable-precise-memory-info \
    --disable-extensions \
    --disable-plugins  \
    --incognito \
    http://localhost:8125
```
* There's a bunch of repos touched, gathered under [ckeditor5 branch](https://github.com/ckeditor/ckeditor5/commits/t/1341):
    - editor-balloon: [PR](https://github.com/ckeditor/ckeditor5-editor-balloon/pull/25)
    - editor-classic: [PR](https://github.com/ckeditor/ckeditor5-editor-classic/pull/78)
    - editor-decoupled: [PR](https://github.com/ckeditor/ckeditor5-editor-decoupled/pull/24)
    - editor-inline: [PR](https://github.com/ckeditor/ckeditor5-editor-inline/pull/44)
    - engine: [PR](https://github.com/ckeditor/ckeditor5-engine/pull/1635)
    - image: [PR](https://github.com/ckeditor/ckeditor5-image/pull/270)
    - link: [PR](https://github.com/ckeditor/ckeditor5-link/pull/212)
    - ~typing: [PR](https://github.com/ckeditor/ckeditor5-typing/pull/177)~ (redundant)
    - ui: [PR](https://github.com/ckeditor/ckeditor5-ui/pull/459)
    - utils: [PR](https://github.com/ckeditor/ckeditor5-utils/pull/270)
    - widget: [PR](https://github.com/ckeditor/ckeditor5-widget/pull/65)

WIP - test are failing.

jodator committed:
Introduce test utils for testing memory leaks with Karma.

jodator committed:
Improve editor destructuring process.

jodator committed:
Add memory leak test retry run if it fails.

jodator committed:
Make memory test fail if memory grows over 0 bytes.

jodator committed:
Remove redundant timeouts and garbage collector calls.

Reinmar commented:
Frankly speaking, this entire code should not be necessary. If we really need all that dereferencing, then it means that something still retains the `editor` reference, which means that there's some issue that those PRs don't fix.

Also, like I said on the TM, I think that manually destroying UI components is also unnecessary. They should be destroyed by their parents or by the plugin which defined the given root UI component.

I may be wrong, but I think we should look into this more closely because I think this is just hiding some lower-level issue that we have.

jodator commented:
> Also, like I said on the TM, I think that manually destroying UI components is also unnecessary. They should be destroyed by their parents or by the plugin which defined the given root UI component.

Yes I they weren't destroyed by the parent Plugin - that's the part of the fix.
The other code was needed mostly because of the `clickOutsideHandler` that attached some listeners to the DOM that no one cleared and some references were still there.

I will review the code to what parts are really needed to fix "static" memory leaks (mostly the `FocusHandler` or similar dereferences but almost all of those changes were needed to free up memory after properly destroying the editor - which involves properly dereferencing it in the test (ie manual test).

~ps.: also the PR is not ready as I need to fix some tests thay fails.~ I've added the note as I removed it by mistake.

jodator committed:
Remove redundant variables de-referencing.

oleq commented:
I noticed certain instability when running tests. Locally:

```
  BalloonEditor build
    memory usage
      ✖ should not grow on multiple create/destroy
        Chrome 71.0.3578 (Mac OS X 10.14.2)
      AssertionError: used heap size should not grow: expected 6141640 to be at most 0

      + expected - actual

      -6141640
      +0

    at Promise.resolve.then.then.then.then.then.then.then.then.then.then.then.then.then.then (webpack:///packages/ckeditor5-core/tests/_utils/memory.js:67 <- build/.automated-tests/entry-point.js:43065:74)


  DecoupledEditor build
    memory usage
      ✖ should not grow on multiple create/destroy
        Chrome 71.0.3578 (Mac OS X 10.14.2)
      Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```

[and on CI](https://travis-ci.org/ckeditor/ckeditor5/jobs/480780291)

```
Chrome 71.0.3578 (Linux 0.0.0) ClassicEditor build memory usage should not grow on multiple create/destroy FAILED
	AssertionError: used heap size should not grow: expected 2766784 to be at most 0
	    at Promise.resolve.then.then.then.then.then.then.then.then.then.then.then.then.then.then (webpack:///./packages/ckeditor5-core/tests/_utils/memory.js?:70:74)
..........17 01 2019 09:29:27.839:WARN [web-server]: 404: /manual/sample.jpg
..............17 01 2019 09:29:30.514:WARN [web-server]: 404: /manual/sample.jpg
........
Chrome 71.0.3578 (Linux 0.0.0) DecoupledEditor build memory usage should not grow on multiple create/destroy FAILED
	Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```

note that a different editor type got memory bloated.

jodator commented:
It is stable-ish AFAICT: 
![selection_163](https://user-images.githubusercontent.com/247363/51314582-61add980-1a50-11e9-8185-e3c5cbfe983a.png)

jodator commented:
ps.: It's only the builds which fails so it might be that some plugin which I didn't test in `memory` manual tests is causing this. I'll try to debug it though.

Other solution is to close an eye for builds and remove those tests ;)

jodator committed:
Increase timeout in memory tests.

jodator commented:
@oleq: after some debugging and also talking with @mlewand it seems that it might beneficial to add some timeouts after `window.gc()` calls. I've also updated the timeout for whole test as it looks like one of the tests is taking longer to complete (`DecoupledEditor` tests).

Could you test current state of branch constellation (sorry for messy commits - I needed to trigger CI couple of times) on you machine? Current configuration will run every test 10 times so it could take a while to complete. I'll remove unnecessary `it.only()` and `for` loop from the final PR. Right not I'd like to check current setup for stability.

```sh
yarn test -cws --files=editor-*,build-*
```

jodator commented:
@oleq: It looks like the CI is [stable enough](https://travis-ci.org/ckeditor/ckeditor5/builds/482366971#L909). The only one fails is test timeout which might be due to "heavy" usage of `window.gc()` during tests (the same tests is repeated 10 times in that run) and IMHO we could live with it. If it the test pass on your local machine I will cleanup git history and we could merge it finally :)

oleq commented:
LGTM

jodator committed:
Make collecting memory statistic a promise.

jodator committed:
Extract magic numbers to constants.

jodator committed:
Update docs for memory usage test functions.

jodator committed:
Merge branch 'master' into t/ckeditor5/1341

jodator commented:
@oleq I'm confident enough that current setup will not blow up after merging it.

The CI build [is :green_heart:](https://travis-ci.org/ckeditor/ckeditor5/builds/482802589?utm_source=github_status&utm_medium=notification).
